src/components/MainSidebar/ProgramDimensionsPanel/ProgramDimensionsPanel.js

```diff
@@ -0,0 +1,190 @@
+import { useDataQuery } from '@dhis2/app-runtime'
+import i18n from '@dhis2/d2-i18n'
+import { NoticeBox, CenteredContent, CircularLoader } from '@dhis2/ui'
+import cx from 'classnames'
+import PropTypes from 'prop-types'
+import React, { useEffect, useState } from 'react'
+import { useSelector, useDispatch } from 'react-redux'
+import {
+    tSetUiProgram,
+    acUpdateUiProgramStage,
+    acClearUiProgram,
+} from '../../../actions/ui.js'
+import { useDebounce } from '../../../modules/utils.js'
+import {
+    sGetUiInputType,
+    sGetUiProgramId,
+    sGetUiProgramStage,
+} from '../../../reducers/ui.js'
+import { INPUT_TYPES } from '../InputPanel/index.js'
+import {
+    ProgramDimensionsFilter,
+    DIMENSION_TYPES,
+} from './ProgramDimensionsFilter.js'
+import { ProgramDimensionsList } from './ProgramDimensionsList.js'
+import styles from './ProgramDimensionsPanel.module.css'
+import { ProgramSelect } from './ProgramSelect.js'
+
+const PROGRAM_TYPES = {
+    WITHOUT_REGISTRATION: 'WITHOUT_REGISTRATION',
+    WITH_REGISTRATION: 'WITH_REGISTRATION',
+}
+const query = {
+    programs: {
+        resource: 'programs',
+        params: {
+            fields: [
+                'id',
+                'displayName',
+                'enrollmentDateLabel',
+                'incidentDateLabel',
+                'programType',
+                'programStages[id,displayName]',
+            ],
+            paging: false,
+        },
+    },
+}
+
+const ProgramDimensionsPanel = ({ visible }) => {
+    const dispatch = useDispatch()
+    const inputType = useSelector(sGetUiInputType)
+    const selectedProgramId = useSelector(sGetUiProgramId)
+    const selectedStageId = useSelector(sGetUiProgramStage)
+    const setSelectedProgramId = (id) => dispatch(tSetUiProgram(id))
+    const { fetching, error, data, refetch, called } = useDataQuery(query, {
+        lazy: true,
+    })
+    const [searchTerm, setSearchTerm] = useState('')
+    const [dimensionType, setDimensionType] = useState(DIMENSION_TYPES.ALL)
+    const debouncedSearchTerm = useDebounce(searchTerm, 350)
```

Comment:
`200` is used here https://github.com/dhis2/event-reports-app/blob/26c810f6fc866be4f6d2e83c684471766db0ca3f/src/components/Dialogs/Conditions/OptionSetCondition.js#L100, no idea which is correct, but probably good to be consistent. @cooper-joe ?

Reply:
If the response is not expected to be too "expensive" then a shorter debounce is preferable, but there's flexibility in each use case. Anything between `200` and `500` is acceptable though. If it doesn't really matter resource-wise then let's stick with a consistent `200`.

Reply:
I think @martinkrulltott makes a good point here about using a consistent debounce time for text inputs across components (and probably the platform as a whole really). But what should this value be?

**TL;DR:** 
I think somewhere between 500-750ms is probably a sensible debounce time for inputs which trigger api requests.

---

**Full motivation**
I find 200ms way too short to debounce an input which triggers an api request. Average time between between key strokes varies between users, contexts and probably characters-location on the keyboard, but the lowest estimate I've found so far mentions 167ms. Our 200ms is way to close to that for this type of use-case. When you do work in the client only, it might make sense to have short debounce time, purely to ensure the UI does not become blocked. But for this type of thing you actually want to choose a debounce time which is a good indication the user has actually finished typing, so you avoid sending too many requests with incomplete search terms.

So, really what we want here is to wait as long as possible, but not so long the user starts thinking the app has become unresponsive. Here's some info about that:
> A delay of less than 100 milliseconds feels instant to a user, but a delay between 100 and 300 milliseconds is perceptible. A delay between 300 and 1,000 milliseconds makes the user feel like a machine is working, but if the delay is above 1,000 milliseconds, your user will likely start to mentally context-switch.
_[Source](https://designingforperformance.com/performance-is-ux/#:~:text=A%20delay%20of%20less%20than,start%20to%20mentally%20context%2Dswitch.)_

So this leads me to the following conclusions:
- The upper bound for a debounce time is 1000ms, if we exceed that the user will become distracted
- The lower bound is a bit more guess work, but I would multiplying the average by 2.5 would be enough to accommodate for differences between users/context/characters, so that `~420` ms.
- So I guess the sweet spot is probably a safe range above the lower and below the upper bound, i.e. *500-750ms*

---
**For reference**
Just to show that choosing a debounce time which is too short can cause really bad issues, please see issue [DHIS2-10628](https://jira.dhis2.org/browse/DHIS2-10628), which was a performance issue (as such, perhaps a bit of an extreme example). The debounce time there was `300` ms to begin with, but because we were getting servers freezing due to too many requests to the users endpoint, we ended up increasing the value to `1200` ms. Now I have to admit that this is a super high value and when using the app, the search does feel a bit sluggish. We have probably taken this a bit too far, but at least it seems to have solved the performance issue....